allennlp/training/metrics/spearman_correlation.py

```diff
@@ -58,17 +63,40 @@ def __call__(
             self.total_predictions = torch.cat((self.total_predictions, predictions), 0)
             self.total_gold_labels = torch.cat((self.total_gold_labels, gold_labels), 0)
 
+        if is_distributed():
+            try:
+                # The following logic will only work if the batches are of equal length.
+                world_size = dist.get_world_size()
+                device = gold_labels.device
+                _total_predictions = [
+                    torch.zeros(self.total_predictions.shape, device=device)
+                    for i in range(world_size)
+                ]
+                _total_gold_labels = [
+                    torch.zeros(self.total_gold_labels.shape, device=device)
+                    for i in range(world_size)
+                ]
+                dist.all_gather(_total_predictions, self.total_predictions)
+                dist.all_gather(_total_gold_labels, self.total_gold_labels)
+                self.total_predictions = torch.cat(_total_predictions, dim=0)
+                self.total_gold_labels = torch.cat(_total_gold_labels, dim=0)
+            except Exception:
+                logger.warning(
+                    "Distributed aggregation for SpearmanCorrelation is currently not supported "
+                    "for batches of unequal length."
+                )

```

Comment:
I am happier with the "equal-length batches" restriction if we at least fail properly when the assumption is violated.

Reply:
Oh, I see this is just a warning. I'm thinking we should throw an exception. Otherwise people will run this stuff, miss the warning, copy the number into their paper and publish it, and then get called out and it's our fault. @matt-gardner, do you have thoughts about this?

How hard is it to get rid of the equal-length limitation completely?

Reply:
That makes sense. For backwards compatibility, I was returning what was already being returned. But perhaps, failing here, and making the model explicitly take care of handling the failure is a better way.